llvm/lib/Transforms/Vectorize/SandboxVectorizer/Region.cpp

```diff
@@ -42,4 +54,27 @@ void Region::dump() const {
 }
 #endif // NDEBUG
 
+SmallVector<std::unique_ptr<Region>> Region::createRegionsFromMD(Function &F) {
+  SmallVector<std::unique_ptr<Region>> Regions;
+  DenseMap<MDNode *, Region *> MDNToRegion;
+  auto &Ctx = F.getContext();
+  for (BasicBlock &BB : F) {
+    for (Instruction &Inst : BB) {
+      if (auto *MDN = cast<llvm::Instruction>(Inst.Val)->getMetadata(MDKind)) {
+        Region *R = nullptr;
+        auto It = MDNToRegion.find(MDN);
```

Comment:
Perhaps move this code into a lambda?
```
auto GetOrCreateNode = [&MDNToRegion](MDNode *MDN, std::unique_ptr<Region> &Regions) {
  Region *R = nullptr;
  auto It = MDNToRegion.find(MDN);
  if (It == MDNToRegion.end()) {
     Regions.push_back(std::make_unique<Region>(Ctx));
     R = Regions.back().get();
     MDNToRegion[MDN] = R;
  } else {
     R = It->second;
  }
  return R;
};
Region *R = GetOrCreateNode(MDN, Regions);
R->add(&Inst);
```

Reply:
I don't think the lambda wins us any clarity here. It needs to access both `Regions` and `MDNToRegion` from the surrounding scope, so it's not very self-contained anyway, and this block is just 10 lines of code.